### DISCUSSION TITLE
LineStyle and native in version 8

### DISCUSSION THREAD

**Discussion by Pe443** [pixijs/pixijs#10350] [v8 General]
Hi, I switch to the new version V8 but I have the following question:
What is the new syntax for PIXI.LineStyle and native = true so that lines are always 1 pixel, regardless of whether the object is scaled or not?
I can not find any documentation what I have to do to set this in StrokeStyle

THX

**Comment by bigtimebuddy** [comment]
Native line drawing has not be implemented yet in v8. cc @GoodBoyDigital

**Reply by Pe443** [reply]
Ok, but it will because I do a lot of scaling and want "sharp" lines ;) ? 
Thx

**Reply by bigtimebuddy** [reply]
If you need this feature, I'd suggest staying on v7 for the time being.

**Reply by Pe443** [reply]
Than I wait and thanks for the quick help !!!!

**Reply by joergplewe** [reply]
I had the same issue, but found that "native" lines actually are no perfect solution, bc. their width varies with `window.devicePixelRatio`.

**Comment by elby22** [comment]
@bigtimebuddy If I understand correctly, the native lines feature might not be portable, and has limitations on some browser implementations (`gl.lineWidth()` may not support values > 1). For scaled stroke with configurable line widths, we may be able to implement this as a feature of the stroke style, by expanding the lines in the vertex shader as detailed here: [Drawing Lines is Hard](https://mattdesl.svbtle.com/drawing-lines-is-hard#expanding-in-a-vertex-shader_2). 

Note: I'm not too familiar with WebGPU yet, or how you would do this in canvas, but with SVGs in HTML5, you could use ["vector-effect: non-scaling-stroke"](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/vector-effect)
For WebGL, to do this with batch support it looks like we would need to modify the default vertex shader and add additional attributes for the line normals and miter. Then we can scale the outer vertices of the stroke with the `worldTransformMatrix` to keep the thickness consistent, instead of triangulating the stroke. We can probably refactor `src/scene/graphics/shared/buildCommands/buildLine.ts` to achieve both approaches.

**Comment by Pe443** [comment]
This would be nice because actually I use Pixi-V7 and it works great.

**Comment by Flippsor** [comment]
Would be great to have this feature again!

Especially useful for creating CAD like applications where the line width is always 1 (independent of scale).

**Comment by Pe443** [comment]
This is right, I use it also for CAD-drawings and dimensioning where I need always 1-pixel-lines

**Comment by joergplewe** [comment]
This is dangerous esp. for CAD apps, bc. the actual width of a line on screen depends on `window.devicePixelRatio`. 
Means, on very high resolution displays like Apple Retina lines can be become hardly visible. Might look good on a Dell notebook, but renders unusable on a Mac.
Been there :).

**Comment by 505598397** [comment]
I currently have a project drawing CAD, I am using pixijs8, it is great, I have done a lot of work on this basis. But now the biggest problem is, if I can't use the native line. how can I keep the width of the line at 1 even though the stage is scaled

**Reply by joergplewe** [reply]
What I do is - after zooming - I take the current scale and repaint all lines with a counter-scaled width.
I also do that with icons or labels that have to maintain screensize in scaled/zoomed scenegraphs.

I know this is cumbersome and a performance risk, but nobody showed me a better way so far.

cc @GoodBoyDigital ??

**Comment by lunarraid** [comment]
V8 now uses `pixelLine: true` to accomplish this: https://pixijs.com/8.x/guides/components/graphics-pixel-line

**Reply by joergplewe** [reply]
Good to hear the feature is available again.
Since the line always has native 1pixel (in contrast to CSS 1px) and does not respect `window.devicePixelRatio`, lines become hardly visible on higher resolution displays (`window.devicePixelRatio >= 2`). Next generation displays may make them invisible at all.
So the feature is cool on one hand but useless on the other :).
